### PR TITLE
Refine netlink crate to be reused by other projects

### PR DESCRIPTION
**Pull request by jiangliu**
The netlink crate could be implemented as a generic library and may be reused by other projects.

How can it be improved?

get rid of dependency on the rustjail crate
split Kata specific code out of generic code
reduce unsafe code
reduce the dependency chain. By default it has dependency on about 70 crates.
define features to conditionally enable features.

#241

### CODE REVIEW

File: src/agent/netlink/src/agent_handler.rs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+                // scope
+                rte.scope = rtm.rtm_scope as u32;
+
+                // oif
+                t = attrs[RTA_OIF as usize];
+                if !t.is_null() {
+                    let data = &*(RTA_DATA!(t) as *const i32);
+                    assert_eq!(RTA_PAYLOAD!(t), 4);
+
+                    /*
```
Comment (jodh-intel) [inline src/agent/netlink/src/agent_handler.rs:347]: Can you remove this commented out code? We can always dig it out of `git` later if we need it :smile:

File: src/agent/netlink/src/agent_handler.rs [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+            nlh.addattr32(IFLA_MTU, iface.mtu as u32);
+
+            // if str is null terminated, use addattr_var.
+            // otherwise, use addattr_str
+            nlh.addattr_var(
+                IFLA_IFNAME,
+                iface.name.as_ptr() as *const u8,
+                iface.name.len(),
+            );
+            // addattr_str(nlh, IFLA_IFNAME, iface.name.as_str());
```
Comment (jodh-intel) [inline src/agent/netlink/src/agent_handler.rs:80]: Can this be deleted?

File: src/agent/netlink/src/agent_handler.rs [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        }
+
+        self.rtnl_talk(v.as_mut_slice(), false)?;
+
+        // TODO: why the result is ignored here?
+        let _ = self.set_link_status(&ifinfo, true);
+
+        // test remove this link
+        // let _ = self.remove_interface(iface)?;
```
Comment (jodh-intel) [inline src/agent/netlink/src/agent_handler.rs:89]: Can these 2 lines be deleted?

File: src/agent/netlink/src/lib.rs [hunk 4]
```diff
@@ -3,49 +3,62 @@
...
+//! - all pointers/buffers generated by kernel are valid
+//! - all offsets are within legal range
+//! - all fields have been correctly aligned
 
-mod neigh;
+#![allow(non_camel_case_types)]
+//#![allow(non_upper_case_globals)]
+//#![allow(dead_code)]
+//#![allow(unused_parens)]
+//#![allow(unused_unsafe)]
```
Comment (jodh-intel) [inline src/agent/netlink/src/lib.rs:17]: Can you delete these commented out lines?

File: src/agent/Cargo.toml [hunk 5]
```diff
@@ -9,7 +9,7 @@ oci = { path = "oci" }
 logging = { path = "logging" }
 rustjail = { path = "rustjail" }
 protocols = { path = "protocols" }
-netlink = { path = "netlink" }
+netlink = { path = "netlink", features = ["with-log", "with-agent-handler"] }
```
Comment (jodh-intel) [inline src/agent/Cargo.toml:12]: Nice!

File: src/agent/netlink/src/agent_handler.rs [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+            (Some(src), mask)
+        };
+
+        let gateway = if r.gateway.is_empty() {
+            None
+        } else {
+            Some(parser::parse_ip_addr(r.gateway.as_str())?)
+        };
+
+        /*
```
Comment (jodh-intel) [inline src/agent/netlink/src/agent_handler.rs:474]: Nit: One more block of commented out code to delete :smile:

Comment (jiangliu) [reply src/agent/netlink/src/agent_handler.rs]: Add one more commit to fix this, and also clean up all "cargo clippy" warnings.

**Pull request merged**